arangod/Scheduler/WorkStealingThreadPool.cpp

```diff
@@ -0,0 +1,451 @@
+
+namespace {
+void incCounter(metrics::Counter* cnt, uint64_t delta = 1) {
+  if (cnt) {
+    *cnt += delta;
+  }
+}
+
+auto calculateCoprimes(std::size_t n) -> std::vector<std::size_t> {
+  TRI_ASSERT(n > 0);
+  if (n <= 2) {
+    return {1};
+  }
+  if (n == 3) {
+    return {1, 2};
+  }
+
+  std::vector<std::size_t> result;
+  result.push_back(1);
+  result.push_back(n - 1);
+
+  for (std::size_t i = 2; i <= n / 2; ++i) {
+    if (std::gcd(i, n) == 1) {
+      result.push_back(i);
+      result.push_back(n - i);
+    }
+  }
+  return result;
+}
+}  // namespace
+
+struct WorkStealingThreadPool::ThreadState {
+  ThreadState(std::size_t id, std::size_t stepSize,
+              WorkStealingThreadPool& pool);
+  ~ThreadState();
+
+  void pushBack(std::unique_ptr<WorkItem>&& work) noexcept;
+  void pushFront(std::unique_ptr<WorkItem>&& work) noexcept;
+  void run() noexcept;
+  void signalStop() noexcept;
+  bool belongsToPool(WorkStealingThreadPool const& pool) const noexcept {
+    return &this->pool == &pool;
+  }
+
+  inline static thread_local ThreadState* currentThread = nullptr;
+
+  std::size_t const id;
+
+  std::atomic<std::size_t> prevHint = NoHint;
+
+ private:
+  void wake();
+  void runWork(WorkItem& work) noexcept;
+  auto stealWork() noexcept -> std::unique_ptr<WorkItem>;
+  auto pushMany(WorkItem* item) -> std::unique_ptr<WorkItem>;
+  void goToSleep(std::unique_lock<std::mutex>& guard) noexcept;
+
+  std::atomic<WorkItem*> pushQueue{nullptr};
+  std::atomic<bool> sleeping = false;
+  std::atomic<std::size_t> queueSize = 0;
+  std::size_t const stepSize;
+  WorkStealingThreadPool& pool;
+  std::atomic<std::size_t> stealHint = NoHint;
+
+  alignas(64) std::atomic<bool> stop{false};
+  std::mutex mutex;
+  std::condition_variable cv;
+  std::deque<std::unique_ptr<WorkItem>> _queue;
+  std::mt19937 rng;
+};
+
+WorkStealingThreadPool::ThreadState::ThreadState(std::size_t id,
+                                                 std::size_t stepSize,
+                                                 WorkStealingThreadPool& pool)
+    : id(id), stepSize(stepSize), pool(pool), rng(id) {}
+
+WorkStealingThreadPool::ThreadState::~ThreadState() {
+  auto* item = pushQueue.load(std::memory_order_relaxed);
+  while (item) {
+    auto next = item->next;
+    item->next = nullptr;  // to avoid triggering the assert in the destructor
+    delete item;
+    item = next;
+  }
+}
+
+void WorkStealingThreadPool::ThreadState::wake() {
+  mutex.lock();
+  mutex.unlock();
+  cv.notify_one();
+}
+
+void WorkStealingThreadPool::ThreadState::pushBack(
+    std::unique_ptr<WorkItem>&& work) noexcept {
+  TRI_ASSERT(work != nullptr);
+  TRI_ASSERT(work->next == nullptr);
+  auto p = work.release();
+  p->next = pushQueue.load(std::memory_order_acquire);
+  bool hasWork = p->next != nullptr;
+  // the compare_exchange implicitly reloads the next pointer
+  while (
+      !pushQueue.compare_exchange_weak(p->next, p, std::memory_order_seq_cst)) {
```

Comment:
nit: it seems that this is pushing to the _front_ of the queue, although the method is named `pushBack`. I found that slightly confusing, but it does not need to be changed.

Reply:
The `pushQueue` is basically a stack, so you are right that we are pushing to the front. However, we are not consuming items directly from the `pushQueue`. Instead, any thread that grabs these items pushes them onto it's own queue using `push_front`, effectively reversing the list again.

Reply:
Ok, fine for me.